examples/images_contours_and_fields/pcolormesh_grids.py

```diff
@@ -0,0 +1,141 @@
+"""
+============================
+Pcolormesh grids and shading
+============================
+
+`.axes.Axes.pcolormesh` and `~.axes.Axes.pcolor` have a few options for
+how grids are laid out and the shading between the grid points.
+
+Generally, if *Z* is size *MxN* then the grid *X* and *Y* can be
+specified with either size *M+1 x N+1* or *MxN*, depending on the
+argument for the ``shading`` keyword argument.
+
+"""
+
+import matplotlib
+import matplotlib.pyplot as plt
+from matplotlib.colors import BoundaryNorm
+from matplotlib.ticker import MaxNLocator
+import numpy as np
+
+###############################################################################
+# Flat Shading
+# ------------
+#
+# The grid specification with the least assumptions is ``shading='flat'``
+# and if the grid is one larger than the data in each dimesion, i.e. has size
+# *M+1 x N+1*.  In that case *X* and *Y* sepcify the corners of quadrilaterals
+# that are colored with the values in *Z*:
+#
+#        (X[i+1, j], Y[i+1, j])       (X[i+1, j+1], Y[i+1, j+1])
+#                          +-----+
+#                          |     |
+#                          +-----+
+#        (X[i, j], Y[i, j])       (X[i, j+1], Y[i, j+1])
+#
+# Here we specify the edges of the *3x5* quadirlaterals with *X* and *Y* that
+# are *4x6*.
+np.random.seed(19680801)
+Z = np.random.rand(3, 5)
+x = np.arange(6)  # len = 6
+y = np.arange(4)  # len = 4
+
+fig, ax = plt.subplots()
+ax.pcolormesh(x, y, Z, shading='flat', vmin=Z.min(), vmax=Z.max())
+def _annote(ax, x, y, title):
+    # this all gets repeated below:
+    X, Y = np.meshgrid(x, y)
+    ax.plot(X.flat, Y.flat, 'o', color='m')
```

Comment:
not sure why...  Note I used `scatter` in the example above to use your trick of putting the data values in.  But colouring that way really makes `shading='flat'` look *wrong*, which I don't think is quite what we are after.

Reply:
I agree with your comment about it looking wrong and giving the wrong impression, but you can still call scatter with `color='m'` here. I am just not a fan of calling `plot` to do `scatter` markers without the lines.

My suggestion would be: `ax.scatter(X.flat, Y.flat, color='m')`

Reply:
:) I agree with inconsistencies previously causing some headaches in `scatter`. But, `plot` to me does too much magic under the hood. In this case, you actually get back a list of one object, which is a `Line2D`. For scatter you get a `PathCollection`. I'm used to working with animations and updating the `Collection` objects rather than `Line2D` objects for markers.

As you note though, it is just purely a preference thing for me. I'm not going to hold anything up for this.

Reply:
Thats a side discussion, but can you not animate Line2D objects?

Reply:
You can most definitely animate them, just API differences with the `set_*` calls between `Collection` and `Line2D`. I completely agree I'm taking this down a tangent. You can mark as resolved (I can't figure out how to do that as the original commenter).

Reply:
@greglucas A reason why many of us prefer plot for things like this, when the additional capabilities of scatter are not needed, is that it is much faster.

Reply:
Wow, I actually had no idea there was that much of a speed impact, thanks @efiring! I just did a quick 1M point scatter/plot comparison and scatter was ~3 seconds while plot was ~100 milliseconds. I definitely agree with the decision here now.